{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d6b88720",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "from Bio import SeqIO\n",
    "import subprocess\n",
    "\n",
    "import pandas as pd\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b941f918",
   "metadata": {},
   "source": [
    "## Filter sequencing data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aac3a4a5",
   "metadata": {},
   "source": [
    "To filter sequencing data, we use the software `fastp`, which is executed from the command line. Here, we show an example code that is executed in the command line. We assume that there is a mamba environment installed with the name 'fastp'. Instructions for this environment can be found in the README.md file in the first page of this repository. If `conda` is used instead of `mamba`, simply exchange the code below.\n",
    "\n",
    "Here we have a sample dataset for the mapping step, located in `data/sequencing_reads/mapping/read1_mapping.fastq.gz` for read 1 and `data/sequencing_reads/mapping/read2_mapping.fastq.gz` for read 2, where read 1 contains the sequenced mutated promoter variants, and read 2 contains the corresponding barcodes. We filter the reads for quality and to not contain any unidentified bases, as well as trimming extra bases from the reads (e.g. read 1 is 171 bases, with the promoter variants being in the first 160). The filtered reads are then stored and we can investigate them later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f590e79a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[08:01:43] start to load data of read1 \n",
      "[08:01:43] start to load data of read2 \n",
      "[08:01:44] Read2: loaded 1M reads \n",
      "[08:01:44] Read1: loaded 1M reads \n",
      "[08:01:45] Read2: loaded 2M reads \n",
      "[08:01:45] Read1: loaded 2M reads \n",
      "[08:01:47] Read2: loading completed with 2638 packs \n",
      "[08:01:48] Read1: loading completed with 2638 packs \n",
      "[08:01:48] thread 4 data processing completed \n",
      "[08:01:48] thread 4 finished \n",
      "[08:01:48] thread 3 data processing completed \n",
      "[08:01:48] thread 3 finished \n",
      "[08:01:48] thread 5 data processing completed \n",
      "[08:01:48] thread 5 finished \n",
      "[08:01:48] thread 2 data processing completed \n",
      "[08:01:48] thread 2 finished \n",
      "[08:01:48] thread 6 data processing completed \n",
      "[08:01:48] thread 6 finished \n",
      "[08:01:48] thread 1 data processing completed \n",
      "[08:01:48] thread 1 finished \n",
      "[08:01:48] ../data/filtered_sequencing/mapping/examples/read2_mapping_filtered.fastq.gz writer finished \n",
      "[08:01:49] ../data/filtered_sequencing/mapping/examples/read1_mapping_filtered.fastq.gz writer finished \n",
      "[08:01:49] start to generate reports\n",
      " \n",
      "Read1 before filtering:\n",
      "total reads: 2637127\n",
      "total bases: 421940320\n",
      "Q20 bases: 419143089(99.3371%)\n",
      "Q30 bases: 412495178(97.7615%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 2637127\n",
      "total bases: 52742540\n",
      "Q20 bases: 52316932(99.193%)\n",
      "Q30 bases: 51210723(97.0957%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 2637127\n",
      "total bases: 421940320\n",
      "Q20 bases: 419143089(99.3371%)\n",
      "Q30 bases: 412495178(97.7615%)\n",
      "\n",
      "Read2 after filtering:\n",
      "total reads: 2637127\n",
      "total bases: 52742540\n",
      "Q20 bases: 52316932(99.193%)\n",
      "Q30 bases: 51210723(97.0957%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 5274254\n",
      "reads failed due to low quality: 0\n",
      "reads failed due to too many N: 0\n",
      "reads with adapter trimmed: 0\n",
      "bases trimmed due to adapters: 0\n",
      "\n",
      "Duplication rate: 3.05268%\n",
      "\n",
      "Insert size peak (evaluated by paired-end reads): 160\n",
      "\n",
      "JSON report: fastp.json\n",
      "HTML report: fastp.html\n",
      "\n",
      "fastp --in1 ../data/sequencing_reads/mapping/examples/read1_mapping.fastq.gz --in2 ../data/sequencing_reads/mapping/examples/read2_mapping.fastq.gz --out1 ../data/filtered_sequencing/mapping/examples/read1_mapping_filtered.fastq.gz --out2 ../data/filtered_sequencing/mapping/examples/read2_mapping_filtered.fastq.gz --verbose --disable_length_filtering --thread 6 -q 20 --n_base_limit 0 --unqualified_percent_limit 10 \n",
      "fastp v0.23.4, time used: 7 seconds\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['./filtering_mapping_example.sh'], returncode=0)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtering_code = \"\"\"#!/usr/bin/env bash\n",
    "READ1=\"../data/sequencing_reads/mapping/examples/read1_mapping.fastq.gz\"\n",
    "READ2=\"../data/sequencing_reads/mapping/examples/read2_mapping.fastq.gz\"\n",
    "OUT1=\"../data/filtered_sequencing/mapping/examples/read1_mapping_filtered.fastq.gz\"\n",
    "OUT2=\"../data/filtered_sequencing/mapping/examples/read2_mapping_filtered.fastq.gz\"\n",
    "\n",
    "mamba run -n fastp fastp \\\n",
    "       --in1 \"$READ1\" --in2 \"$READ2\" \\\n",
    "       --out1 \"$OUT1\" --out2 \"$OUT2\" \\\n",
    "       --verbose --disable_length_filtering \\\n",
    "       --thread 6 -q 20 --n_base_limit 0 --unqualified_percent_limit 10\n",
    "\"\"\"\n",
    "\n",
    "with open(\"filtering_mapping_example.sh\", \"w\") as f:\n",
    "    f.write(filtering_code)\n",
    "\n",
    "subprocess.run(['chmod', '+x', './filtering_mapping_example.sh'], text=True)\n",
    "subprocess.run(['./filtering_mapping_example.sh'], check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7f14029",
   "metadata": {},
   "source": [
    "There is also a sample dataset for barcode counts, located at `/data/sequencing_reads/barcodes/gc-rep_DNA.fastq.gz` and `/data/sequencing_reads/barcodes/gc-rep_DNA.fastq.gz`. For each growth condition and replicate, we have have a pair of files for counts from RNA and DNA sequencing. Both files are processed in the same step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "28df80dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Detecting adapter sequence for read1...\n",
      "No adapter detected for read1\n",
      "\n",
      "[17:29:32] start to load data \n",
      "[17:29:32] Loading completed with 11 packs \n",
      "[17:29:32] thread 1 data processing completed \n",
      "[17:29:32] thread 1 finished \n",
      "[17:29:32] thread 5 data processing completed \n",
      "[17:29:32] thread 5 finished \n",
      "[17:29:32] thread 2 data processing completed \n",
      "[17:29:32] thread 2 finished \n",
      "[17:29:32] thread 3 data processing completed \n",
      "[17:29:32] thread 3 finished \n",
      "[17:29:32] thread 6 data processing completed \n",
      "[17:29:32] thread 6 finished \n",
      "[17:29:32] thread 4 data processing completed \n",
      "[17:29:32] thread 4 finished \n",
      "[17:29:32] ../data/filtered_sequencing/barcodes/examples/gc-rep_DNA.fastq.gz writer finished \n",
      "[17:29:32] start to generate reports\n",
      " \n",
      "Read1 before filtering:\n",
      "total reads: 10093\n",
      "total bases: 262418\n",
      "Q20 bases: 259158(98.7577%)\n",
      "Q30 bases: 258075(98.345%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 10093\n",
      "total bases: 201860\n",
      "Q20 bases: 198724(98.4464%)\n",
      "Q30 bases: 197646(97.9124%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 10093\n",
      "reads failed due to low quality: 0\n",
      "reads failed due to too many N: 0\n",
      "reads with adapter trimmed: 0\n",
      "bases trimmed due to adapters: 0\n",
      "\n",
      "Duplication rate (may be overestimated since this is SE data): 34.5289%\n",
      "\n",
      "JSON report: fastp.json\n",
      "HTML report: fastp.html\n",
      "\n",
      "fastp --in1 ../data/sequencing_reads/barcodes/examples/gc-rep_DNA.fastq.gz --out1 ../data/filtered_sequencing/barcodes/examples/gc-rep_DNA.fastq.gz --trim_tail1 6 --verbose --disable_length_filtering --thread 6 --n_base_limit 0 \n",
      "fastp v0.23.4, time used: 0 seconds\n",
      "Detecting adapter sequence for read1...\n",
      "No adapter detected for read1\n",
      "\n",
      "[17:29:32] start to load data \n",
      "[17:29:32] Loading completed with 15 packs \n",
      "[17:29:32] thread 4 data processing completed \n",
      "[17:29:32] thread 4 finished \n",
      "[17:29:32] thread 6 data processing completed \n",
      "[17:29:32] thread 6 finished \n",
      "[17:29:32] thread 5 data processing completed \n",
      "[17:29:32] thread 5 finished \n",
      "[17:29:32] thread 3 data processing completed \n",
      "[17:29:32] thread 3 finished \n",
      "[17:29:32] thread 2 data processing completed \n",
      "[17:29:32] thread 2 finished \n",
      "[17:29:32] thread 1 data processing completed \n",
      "[17:29:32] thread 1 finished \n",
      "[17:29:32] ../data/filtered_sequencing/barcodes/examples/gc-rep_RNA.fastq.gz writer finished \n",
      "[17:29:32] start to generate reports\n",
      " \n",
      "Read1 before filtering:\n",
      "total reads: 14417\n",
      "total bases: 374842\n",
      "Q20 bases: 369846(98.6672%)\n",
      "Q30 bases: 368101(98.2016%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 14417\n",
      "total bases: 288340\n",
      "Q20 bases: 283512(98.3256%)\n",
      "Q30 bases: 281776(97.7235%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 14417\n",
      "reads failed due to low quality: 0\n",
      "reads failed due to too many N: 0\n",
      "reads with adapter trimmed: 0\n",
      "bases trimmed due to adapters: 0\n",
      "\n",
      "Duplication rate (may be overestimated since this is SE data): 68.8215%\n",
      "\n",
      "JSON report: fastp.json\n",
      "HTML report: fastp.html\n",
      "\n",
      "fastp --in1 ../data/sequencing_reads/barcodes/examples/gc-rep_RNA.fastq.gz --out1 ../data/filtered_sequencing/barcodes/examples/gc-rep_RNA.fastq.gz --trim_tail1 6 --verbose --disable_length_filtering --thread 6 --n_base_limit 0 \n",
      "fastp v0.23.4, time used: 0 seconds\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['./filtering_barcode_example.sh'], returncode=0)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtering_code = \"\"\"#!/usr/bin/env bash\n",
    "# Find read\n",
    "READ_DNA=\"../data/sequencing_reads/barcodes/examples/gc-rep_DNA.fastq.gz\"\n",
    "READ_RNA=\"../data/sequencing_reads/barcodes/examples/gc-rep_RNA.fastq.gz\"\n",
    "\n",
    "# Define output file paths\n",
    "OUT_DNA=\"../data/filtered_sequencing/barcodes/examples/gc-rep_DNA.fastq.gz\"\n",
    "OUT_RNA=\"../data/filtered_sequencing/barcodes/examples/gc-rep_RNA.fastq.gz\"\n",
    "\n",
    "# Define string to be ran on the termina\n",
    "mamba run -n fastp fastp --in1 $READ_DNA --out1 $OUT_DNA --trim_tail1 '6'  --verbose --disable_length_filtering --thread '6' --n_base_limit '0'\n",
    "mamba run -n fastp fastp --in1 $READ_RNA --out1 $OUT_RNA --trim_tail1 '6'  --verbose --disable_length_filtering --thread '6' --n_base_limit '0'\n",
    "\"\"\"\n",
    "\n",
    "with open(\"filtering_barcode_example.sh\", \"w\") as f:\n",
    "    f.write(filtering_code)\n",
    "\n",
    "subprocess.run(['chmod', '+x', './filtering_barcode_example.sh'], text=True)\n",
    "subprocess.run(['./filtering_barcode_example.sh'], check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "829ce6d2",
   "metadata": {},
   "source": [
    "## Extracting promoter variants and barcodes from sequencing files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85f0af86",
   "metadata": {},
   "source": [
    "First we have to import the sequencing data, which is in the compressed `gzip` format. We decompress the data and import the `fastq` files, then extract the sequence from each record."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "e51f5c26",
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_fastq_gz(filename):\n",
    "    \"\"\"\n",
    "    Imports a gzipped FASTQ file and returns a generator of SeqRecord objects.\n",
    "\n",
    "    Args:\n",
    "        filename (str): The path to the gzipped FASTQ file.\n",
    "\n",
    "    Yields:\n",
    "        SeqRecord: A SeqRecord object representing a single read.\n",
    "    \"\"\"\n",
    "    with gzip.open(filename, \"rt\") as handle:\n",
    "        for record in SeqIO.parse(handle, \"fastq\"):\n",
    "            yield record\n",
    "\n",
    "\n",
    "# Get promoters\n",
    "file_path = \"../data/filtered_sequencing/mapping/examples/read1_mapping_filtered.fastq.gz\"\n",
    "\n",
    "#file_path = \"../data/sequencing_reads/mapping/examples/100_R1.barcode.fastq.gz\"\n",
    "\n",
    "records = import_fastq_gz(file_path)\n",
    "\n",
    "promoters = [str(r.seq) for r in records]\n",
    "\n",
    "# Get barcodes\n",
    "file_path = \"../data/filtered_sequencing/mapping/examples/read2_mapping_filtered.fastq.gz\"\n",
    "#file_path = \"../data/sequencing_reads/mapping/examples/100_R2.barcode.fastq.gz\"\n",
    "\n",
    "records = import_fastq_gz(file_path)\n",
    "barcodes = [str(r.seq) for r in records]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dc837af",
   "metadata": {},
   "source": [
    "Now we count how often each unique pair of barcode and promoter variant is observed in the dataset. This step is very simple using `Pandas`, where we create a `DataFrame` and then count how ofted a row occurs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "5a0fdb91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>promoter_variant</th>\n",
       "      <th>barcode</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGCAATTTCGGCACGAATTTTGACGTATTTAGTGCATAGCTGAGTA...</td>\n",
       "      <td>TACTCGCGGGAGAATAAATG</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGCAATTTCTGCACGAATCTTGACGTATTTAGTGCATAGTTGAGTA...</td>\n",
       "      <td>GAAACGCGTCTGAGTTTTTG</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CGCAATTTCGGCGGGAATTTTGACGTGTTTAGTGCAAAGTTCAGTA...</td>\n",
       "      <td>CCTGATGGAGGTTAAAGAGC</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGCAATTTCCGCACGAATTTTGACGTTTTTAGGGCATTGTTGAGTA...</td>\n",
       "      <td>TACTCGCGGGAGAATAAATG</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGCAATCTCGGCACGAATTTTGACGTATTTGGTGCATAGTTGAGTA...</td>\n",
       "      <td>GTTCTATTTAAAGTGCGCTT</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    promoter_variant               barcode  \\\n",
       "0  CGCAATTTCGGCACGAATTTTGACGTATTTAGTGCATAGCTGAGTA...  TACTCGCGGGAGAATAAATG   \n",
       "1  CGCAATTTCTGCACGAATCTTGACGTATTTAGTGCATAGTTGAGTA...  GAAACGCGTCTGAGTTTTTG   \n",
       "2  CGCAATTTCGGCGGGAATTTTGACGTGTTTAGTGCAAAGTTCAGTA...  CCTGATGGAGGTTAAAGAGC   \n",
       "3  CGCAATTTCCGCACGAATTTTGACGTTTTTAGGGCATTGTTGAGTA...  TACTCGCGGGAGAATAAATG   \n",
       "4  CGCAATCTCGGCACGAATTTTGACGTATTTGGTGCATAGTTGAGTA...  GTTCTATTTAAAGTGCGCTT   \n",
       "\n",
       "   count  \n",
       "0      5  \n",
       "1      5  \n",
       "2      5  \n",
       "3      5  \n",
       "4      5  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_counts = (\n",
    "    pd.DataFrame({\"promoter_variant\": promoters, \"barcode\": barcodes})\n",
    "      .value_counts()                    \n",
    "      .reset_index(name=\"count\")       \n",
    ")\n",
    "\n",
    "df_counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "0ad5fc5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_counts['promoter_variant'].values[0] == 'AACAATTTCCGCACGCATTTTGACGTATTTAGTGCATTGTGGAGTATGGATCACAGTTTGTGTTTCGTCCAAATATTACTGTTTTTTTATAAAGGAAACTTCTATAATATCACTTTTCGCAATGTGTTATGCGGGGGCGGCATCCTTACCCGGCGCACTA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "37bbde72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(76)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_counts[df_counts['barcode'] == 'ACCAGTAGATACGACGTGCA']['count'].sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fed4fdcf",
   "metadata": {},
   "source": [
    "## Identify Promoter Variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "019fd0ca",
   "metadata": {},
   "source": [
    "Now that we the promoter variants and barcodes in hand, we have to identify which promoter a sequence belongs to. Here, we use a simplified version where we only look for perfect matches with one sequence in the list of created mutants. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e7e6e9fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>promoter_variant</th>\n",
       "      <th>barcode</th>\n",
       "      <th>count</th>\n",
       "      <th>promoter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGCAATTTCGGCACGAATTTTGACGTATTTAGTGCATAGCTGAGTA...</td>\n",
       "      <td>TACTCGCGGGAGAATAAATG</td>\n",
       "      <td>5</td>\n",
       "      <td>tisBp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGCAATTTCTGCACGAATCTTGACGTATTTAGTGCATAGTTGAGTA...</td>\n",
       "      <td>GAAACGCGTCTGAGTTTTTG</td>\n",
       "      <td>5</td>\n",
       "      <td>tisBp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CGCAATTTCGGCGGGAATTTTGACGTGTTTAGTGCAAAGTTCAGTA...</td>\n",
       "      <td>CCTGATGGAGGTTAAAGAGC</td>\n",
       "      <td>5</td>\n",
       "      <td>tisBp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGCAATTTCCGCACGAATTTTGACGTTTTTAGGGCATTGTTGAGTA...</td>\n",
       "      <td>TACTCGCGGGAGAATAAATG</td>\n",
       "      <td>5</td>\n",
       "      <td>tisBp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CGCAATCTCGGCACGAATTTTGACGTATTTGGTGCATAGTTGAGTA...</td>\n",
       "      <td>GTTCTATTTAAAGTGCGCTT</td>\n",
       "      <td>5</td>\n",
       "      <td>tisBp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    promoter_variant               barcode  \\\n",
       "0  CGCAATTTCGGCACGAATTTTGACGTATTTAGTGCATAGCTGAGTA...  TACTCGCGGGAGAATAAATG   \n",
       "1  CGCAATTTCTGCACGAATCTTGACGTATTTAGTGCATAGTTGAGTA...  GAAACGCGTCTGAGTTTTTG   \n",
       "2  CGCAATTTCGGCGGGAATTTTGACGTGTTTAGTGCAAAGTTCAGTA...  CCTGATGGAGGTTAAAGAGC   \n",
       "3  CGCAATTTCCGCACGAATTTTGACGTTTTTAGGGCATTGTTGAGTA...  TACTCGCGGGAGAATAAATG   \n",
       "4  CGCAATCTCGGCACGAATTTTGACGTATTTGGTGCATAGTTGAGTA...  GTTCTATTTAAAGTGCGCTT   \n",
       "\n",
       "   count promoter  \n",
       "0      5    tisBp  \n",
       "1      5    tisBp  \n",
       "2      5    tisBp  \n",
       "3      5    tisBp  \n",
       "4      5    tisBp  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import sequences\n",
    "df_wt = pd.read_csv(\"../data/metadata/promoter_variants.csv\")\n",
    "\n",
    "# keep only needed columns\n",
    "df_wt = df_wt[['sequence', 'promoter']]\n",
    "\n",
    "# remove overhangs from sequences\n",
    "df_wt['sequence'] = [x[26:186] for x in df_wt['sequence']]\n",
    "\n",
    "# merge counts and promoters\n",
    "df_map = df_counts.merge(df_wt, how='left', left_on='promoter_variant', right_on='sequence')\n",
    "\n",
    "# remove promoters without hits\n",
    "df_map.dropna(inplace=True, axis=0)\n",
    "\n",
    "# drop duplicate column\n",
    "df_map.drop(columns=['sequence'], inplace=True)\n",
    "df_map.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5479982c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_map.to_csv('../data/barcode_map/examples/example_map.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3253272d",
   "metadata": {},
   "source": [
    "## Process Barcode counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3caa128d",
   "metadata": {},
   "source": [
    "Now we take a look at the barcode counts. We import the filtered sequencing data and extract the barcodes from the data. Then we count how often a barcode was found in the DNA data and how often it was found in the RNA data. Then we merge the data into a single dataframe which contains the barcode and counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "19a93ad5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>ct_0</th>\n",
       "      <th>ct_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAACACAGATACTCCGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAACGGAAATAAAGCGG</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAATCTATTGCGTACAT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAATGAGGTGATCAGGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAACACGTCAGGCATGCA</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                barcode  ct_0  ct_1\n",
       "0  AAAAAACACAGATACTCCGA   1.0   0.0\n",
       "1  AAAAAACGGAAATAAAGCGG   1.0   0.0\n",
       "2  AAAAAATCTATTGCGTACAT   0.0   1.0\n",
       "3  AAAAAATGAGGTGATCAGGA   1.0   0.0\n",
       "4  AAAAACACGTCAGGCATGCA   2.0   3.0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get dna barcodes\n",
    "file_path = \"../data/filtered_sequencing/barcodes/examples/gc-rep_DNA.fastq.gz\"\n",
    "\n",
    "# extract sequences\n",
    "records = import_fastq_gz(file_path)\n",
    "dna_barcodes = [str(r.seq) for r in records]\n",
    "\n",
    "# build dataframe and count barcodes\n",
    "df_DNA = pd.DataFrame(data={\"barcode\":dna_barcodes}).value_counts().reset_index()\n",
    "df_DNA.rename(columns={\"count\": \"ct_0\"}, inplace=True)\n",
    "df_DNA.to_csv('../data/extracted_barcodes/examples/DNA_collapsed.txt', index=False)\n",
    "\n",
    "# Get rna barcodes\n",
    "file_path = \"../data/filtered_sequencing/barcodes/examples/gc-rep_RNA.fastq.gz\"\n",
    "\n",
    "# extract sequences\n",
    "records = import_fastq_gz(file_path)\n",
    "rna_barcodes = [str(r.seq) for r in records]\n",
    "\n",
    "# build dataframe and count barcodes\n",
    "df_RNA = pd.DataFrame(data={\"barcode\": rna_barcodes}).value_counts().reset_index()\n",
    "df_RNA.rename(columns={\"count\": \"ct_1\"}, inplace=True)\n",
    "df_RNA.to_csv('../data/extracted_barcodes/examples/RNA_collapsed.txt', index=False)\n",
    "\n",
    "# combine count dataframes\n",
    "df_barcodes = df_DNA.merge(df_RNA, how='outer', on='barcode').fillna(0)\n",
    "df_barcodes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8ab3085",
   "metadata": {},
   "source": [
    "Now we identify for each barcode which promoter it belongs to. In this example dataset, we only find barcodes for the *tisB* promoter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "56d9fda4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>ct_0</th>\n",
       "      <th>ct_1</th>\n",
       "      <th>promoter_variant</th>\n",
       "      <th>count</th>\n",
       "      <th>promoter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAACACAGATACTCCGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CGCAATTTCGGGACGGATTTTGACGTAATTAGTGCATAGTTGAGTA...</td>\n",
       "      <td>1</td>\n",
       "      <td>tisBp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAACACAGATACTCCGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CGCAATTTCGGCACGAATTTTGACGTGTTTAGTGAAGAGTTGAGTA...</td>\n",
       "      <td>1</td>\n",
       "      <td>tisBp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAACACAGATACTCCGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CGGTATTTGGGCTCGAGTTGTGACGTATTTAGTGCATAGTTGAGTA...</td>\n",
       "      <td>1</td>\n",
       "      <td>tisBp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAACACAGATACTCCGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TGCAATTTCGGCACAAATTTCGACGTATTGAGTACATAGTTGAGTA...</td>\n",
       "      <td>1</td>\n",
       "      <td>tisBp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAACACAGATACTCCGA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TGCAATTTGGGCACGAAGTTTGACGTCTTTAATGCATAGGTGAGTA...</td>\n",
       "      <td>1</td>\n",
       "      <td>tisBp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                barcode  ct_0  ct_1  \\\n",
       "0  AAAAAACACAGATACTCCGA   1.0   0.0   \n",
       "1  AAAAAACACAGATACTCCGA   1.0   0.0   \n",
       "2  AAAAAACACAGATACTCCGA   1.0   0.0   \n",
       "3  AAAAAACACAGATACTCCGA   1.0   0.0   \n",
       "4  AAAAAACACAGATACTCCGA   1.0   0.0   \n",
       "\n",
       "                                    promoter_variant  count promoter  \n",
       "0  CGCAATTTCGGGACGGATTTTGACGTAATTAGTGCATAGTTGAGTA...      1    tisBp  \n",
       "1  CGCAATTTCGGCACGAATTTTGACGTGTTTAGTGAAGAGTTGAGTA...      1    tisBp  \n",
       "2  CGGTATTTGGGCTCGAGTTGTGACGTATTTAGTGCATAGTTGAGTA...      1    tisBp  \n",
       "3  TGCAATTTCGGCACAAATTTCGACGTATTGAGTACATAGTTGAGTA...      1    tisBp  \n",
       "4  TGCAATTTGGGCACGAAGTTTGACGTCTTTAATGCATAGGTGAGTA...      1    tisBp  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df_barcodes.merge(df_map, how='inner', on='barcode')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0b029066",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "160"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df['promoter_variant'].values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd680e39",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bebi103",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
